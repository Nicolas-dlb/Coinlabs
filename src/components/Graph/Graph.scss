.graph_container {
	.graph {
		position: relative;
		top: 4vh;

		display: flex;
		flex-direction: column;
		align-items: center;
		justify-content: center;
		.recharts-responsive-container {
			position: relative;
			top: 2vh;
		}
		width: 100%;
		.tooltip-top {
			display: flex;
			align-items: center;
			justify-content: space-between;
			svg {
				width: 11px;
			}
		}
		.tooltip-bottom {
			display: flex;
			align-items: center;
			font-size: 21px;
			margin-top: 15px;
			margin-bottom: 2px;
		}
		&::after {
			content: "";
			position: absolute;
			top: -4vh;
			left: 0;
			right: 5.2%;
			bottom: 70.2vh;
			border-radius: 20px;
			-webkit-box-shadow: inset 0px 13px 31px 39px rgba(34, 30, 52, 1);
			-moz-box-shadow: inset 0px 13px 31px 39px rgba(34, 30, 52, 1);
			box-shadow: inset 0px 13px 31px 39px rgba(34, 30, 52, 1);

			pointer-events: none;
			z-index: 1;
		}
	}
}
