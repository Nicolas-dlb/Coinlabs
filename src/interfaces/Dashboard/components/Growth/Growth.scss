.growth {
  width: 93%;
  height: 36vh;
  background-color: var(--third-bg-color);
  border-radius: 20px;
  margin-left: 10px;
  margin-top: 2vh;
  transition: background-color 0.3s ease;
  .tabs {
    display: flex;
    justify-content: space-between;
    align-items: center;
    padding: 16px 22px;

    .tabs_left {
      display: flex;
      justify-content: space-between;
      align-items: center;
      width: 50%;
      span {
        cursor: pointer;
        color: #c4c3c9;
        &:hover {
          color: #fff;
        }
      }
      .btn_growth_active {
        color: var(--primary-font-color);
      }
    }
    .tabs_right {
      display: flex;
      align-items: flex-start;
      cursor: pointer;
      color: #f8f7f9;

      border-radius: 6px;
      padding: 3px 0px;
      font-size: 15px;
      z-index: 10;

      height: 26px;
      transition: all 0.35s ease-in;

      width: 103px;

      .currency-select {
        position: relative;
        top: 3px;
      }

      svg {
        fill: #c8c6d2;
        height: 22px !important;
        position: relative;
        top: -1px !important;
        left: 73px;
        margin-left: 5px;
        width: auto;
        pointer-events: none;
      }
    }
  }

  .graph {
    position: relative;
    top: 0px;
    .grid {
      display: none !important;
    }
    &::after {
      top: -16vh;
      left: 2%;
      right: 2%;
    }
    .recharts-surface {
      top: -7.5vh;
    }
  }
}
@media only screen and (max-height: 650px) {
  .growth {
    .graph {
      top: -50px;
    }
  }
}

@media only screen and (min-width: 992px) {
  .growth {
    width: 99%;
    margin-left: 0;
    height: 44vh;
    .graph_container {
      width: 97%;
      position: relative;
      top: 2vh;
    }
    .graph {
      .recharts-surface {
        position: relative;
        top: 60px;
      }
    }
    .graph::after {
      top: -4vh;
      bottom: -12vh;
      width: 98.2%;
    }
    .tabs_left {
      .btn_growth {
        border-radius: 4px;
        padding: 5px 15px;
        background: #2b263f;
        width: 23px;
        text-align: center;
        font-size: 15px;
        &:hover {
          background: var(--primary-font-color);
          color: #2d2a38 !important;
        }
      }

      .btn_growth_active {
        background: var(--primary-font-color);
        color: #2d2a38 !important;
      }
    }

    .tabs_right {
      background: #2b263f;

      svg {
        fill: #c8c6d2;
        height: 22px !important;
        position: relative;
        top: -1px !important;
        left: 73px;
        margin-left: 5px;
        width: auto;
        pointer-events: none;
      }
    }
  }
}
