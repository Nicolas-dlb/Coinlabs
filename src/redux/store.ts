import { configureStore } from "@reduxjs/toolkit";
import marketReducer from "./reducers/marketSlice";
import walletsReducer from "./reducers/walletsSlice";
import appReducer from "./reducers/appSlice";
import userReducer from "./reducers/userSlice";

export default configureStore({
  reducer: {
    market: marketReducer,
    wallets: walletsReducer,
    app: appReducer,
    user: userReducer,
  },
});
