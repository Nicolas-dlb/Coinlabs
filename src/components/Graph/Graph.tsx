import {
  AreaChart,
  Area,
  XAxis,
  YAxis,
  Tooltip,
  ResponsiveContainer,
} from "recharts";
import "./Graph.scss";
import React, { useEffect, useState } from "react";
import { getNumberFixed, numberWithSpaces } from "utils/utils";
import Grid from "./Grid/Grid";

type GraphProps = {
  data: Array<Object>;
  color: string;
  width: number;
  height: number | undefined;
  tooltip?: true;
  selected?: string;
};

function Graph({
  data,
  color,
  width,
  height,
  tooltip = undefined,
  selected = "",
}: GraphProps) {
  const [number, setNumber] = useState(0);

  useEffect(() => {
    if (selected === "Bitcoin" || selected === "All") {
      setNumber(0);
    } else if (selected === "Ethereum") {
      setNumber(1);
    } else if (selected === "Litecoin") {
      setNumber(2);
    } else if (selected === "Ripple") {
      setNumber(4);
    } else if (selected === "Neo") {
      setNumber(3);
    }
  }, [selected]);

  const CustomTooltip = ({ active, payload, num }: any) => {
    let angle;
    if (payload) {
      angle = payload[0]?.value > 0 ? "140deg" : "215deg";
    }
    const divise =
      num === 0
        ? 1
        : num === 1
        ? 11
        : num === 2
        ? 100
        : num === 4
        ? 5000
        : num === 3
        ? 200
        : 1;
    const price = payload[num]?.value / divise;
    if (active) {
      return (
        <>
          <div className="tooltip-top">
            <p style={{ marginRight: "25px" }}>Balance</p>
            <svg
              style={{ transform: `rotate(${angle})` }}
              xmlns="http://www.w3.org/2000/svg"
              viewBox="0 0 492 492"
              fill="#fff"
            >
              <path d="M464.3 207.4l0.8 0.2H135.9l103.5-103.7c5.1-5.1 7.8-11.9 7.8-19.1 0-7.2-2.8-14-7.8-19.1L223.3 49.5c-5.1-5.1-11.8-7.9-19-7.9 -7.2 0-14 2.8-19 7.8L7.8 226.9C2.8 232 0 238.8 0 246c0 7.2 2.8 14 7.8 19.1l177.4 177.4c5.1 5.1 11.8 7.8 19 7.8 7.2 0 13.9-2.8 19-7.8l16.1-16.1c5.1-5.1 7.8-11.8 7.8-19 0-7.2-2.8-13.6-7.8-18.7L134.7 284.4h330c14.8 0 27.3-12.8 27.3-27.6v-22.8C492 219.2 479.2 207.4 464.3 207.4z" />
            </svg>
            <p style={{ marginLeft: "6px" }}>12%</p>
          </div>
          <div className="tooltip-bottom">
            <p>$</p>
            <p className="tooltip_price" id="tooltip_price">
              {numberWithSpaces(getNumberFixed(price, 2))}
            </p>
          </div>
        </>
      );
    }

    return null;
  };

  const wrapperStyle = {
    borderRadius: "15px",
    background:
      number === 0
        ? "linear-gradient(20deg, rgba(11,171,81,1) 0%, rgba(30,171,155,1) 100%)"
        : number === 1
        ? "linear-gradient(292deg, rgba(105,162,200,1) 0%, rgba(72,140,185,1) 100%)"
        : number === 2
        ? "linear-gradient(233deg, rgba(203,161,128,1) 0%, rgba(219,154,103,1) 100%)"
        : number === 4
        ? "linear-gradient(225deg, rgba(170,134,190,1) 0%, rgba(160,117,185,1) 30%)"
        : "linear-gradient(233deg, rgba(117,174,218,1) 0%, rgba(74,156,192,1) 40%)",
    border: "none",
    color: "#fff",
    padding: "10px",
    paddingTop: "12px",
    BoxShadow: "0px 6px 18px -6px rgba(0,0,0,0.49)",
    WebkitBoxShadow: "0px 6px 18px -6px rgba(0,0,0,0.49)",
    MozBoxShadow: "0px 6px 18px -6px rgba(0,0,0,0.49)",
    alignItems: "space-evenly",
    justifyContent: "center",
    zIndex: 1000,
  };

  return (
    <div className="graph_container">
      <div className="graph">
        <ResponsiveContainer
          width={width || (!width && tooltip ? "90%" : "99%")}
          height={height}
        >
          <AreaChart width={width} height={height} data={data}>
            <defs>
              <linearGradient id="Green" x1="1" y1="0" x2="0" y2="1">
                <stop offset="0%" stopColor="#2aac63" stopOpacity={0.17} />
                <stop offset="100%" stopColor="#2aac63" stopOpacity={0} />
              </linearGradient>
              <linearGradient id="White" x1="0" y1="0" x2="0" y2="1">
                <stop offset="0%" stopColor="#fff" stopOpacity={0} />
                <stop offset="100%" stopColor="#fff" stopOpacity={0} />
              </linearGradient>
              <linearGradient id="Blue" x1="1" y1="0" x2="0" y2="1">
                <stop offset="0%" stopColor="#4689b6" stopOpacity={0.17} />
                <stop offset="100%" stopColor="#4689b6" stopOpacity={0} />
              </linearGradient>
              <linearGradient id="Pink" x1="1" y1="0" x2="0" y2="1">
                <stop offset="0%" stopColor="#a075b9" stopOpacity={0.17} />
                <stop offset="100%" stopColor="#a075b9" stopOpacity={0} />
              </linearGradient>
              <linearGradient id="Orange" x1="1" y1="0" x2="0" y2="1">
                <stop offset="0%" stopColor="#db9a67" stopOpacity={0.17} />
                <stop offset="100%" stopColor="#db9a67" stopOpacity={0} />
              </linearGradient>
              <linearGradient id="Blue2" x1="1" y1="0" x2="0" y2="1">
                <stop offset="0%" stopColor="#4a9cc0" stopOpacity={0.17} />
                <stop offset="100%" stopColor="#4a9cc0" stopOpacity={0} />
              </linearGradient>
            </defs>
            <XAxis
              type="number"
              dataKey="tokens"
              hide
              interval="preserveStartEnd"
              tickMargin={25}
              tickLine
              axisLine={false}
              tickCount={5}
            />
            <YAxis hide tickMargin={25} tickLine axisLine={false} />
            {tooltip && (
              <Tooltip
                content={<CustomTooltip num={number} />}
                wrapperStyle={wrapperStyle}
                itemStyle={{ color: "rgb(11,171,81)" }}
                cursor={{
                  stroke:
                    number === 0
                      ? "rgb(11,171,81)"
                      : number === 1
                      ? "#4689b6"
                      : number === 2
                      ? "#db9a67"
                      : number === 4
                      ? "#a075b9"
                      : "#4a9cc0",
                  strokeWidth: 1.8,
                  strokeDasharray: 10,
                }}
                labelFormatter={(value: any) => `Tokens: ${value}`}
              />
            )}
            <Area
              activeDot={{
                stroke: "#fff",
                fill: "#26ae5f",
                strokeWidth: number === 0 ? 3 : 0,
                r: number === 0 ? 5 : 0,
              }}
              onMouseOver={() => {
                setNumber(0);
              }}
              strokeWidth={
                selected.includes("Bitcoin") ||
                selected.includes("All") ||
                selected === ""
                  ? 1
                  : 0
              }
              type="monotone"
              dataKey="Price"
              stroke={
                color === "Red"
                  ? "#f0487c"
                  : color === "Green"
                  ? "#2aac63"
                  : color === "White"
                  ? "#fff"
                  : "#fff"
              }
              fillOpacity={1}
              fill={
                selected.includes("Bitcoin") ||
                selected.includes("All") ||
                selected === ""
                  ? `url(#${color})`
                  : "none"
              }
            />

            <Area
              activeDot={{
                stroke: "#fff",
                fill: "#4689b6",
                strokeWidth: number === 1 ? 3 : 0,
                r: number === 1 ? 5 : 0,
              }}
              onMouseOver={() => {
                setNumber(1);
              }}
              strokeWidth={
                selected.includes("Ethereum") || selected.includes("All")
                  ? 1
                  : 0
              }
              type="monotone"
              dataKey="ethereum"
              stroke="#4689b6"
              fillOpacity={1}
              fill={
                selected.includes("Ethereum") || selected.includes("All")
                  ? `url(#Blue)`
                  : "none"
              }
            />
            <Area
              activeDot={{
                stroke: "#fff",
                fill: "#db9a67",
                strokeWidth: number === 2 ? 3 : 0,
                r: number === 2 ? 5 : 0,
              }}
              onMouseOver={() => {
                setNumber(2);
              }}
              strokeWidth={
                selected.includes("Litecoin") || selected.includes("All")
                  ? 1
                  : 0
              }
              type="monotone"
              dataKey="litecoin"
              stroke="#db9a67"
              fillOpacity={1}
              fill={
                selected.includes("Litecoin") || selected.includes("All")
                  ? `url(#Orange)`
                  : "none"
              }
            />

            <Area
              activeDot={{
                stroke: "#fff",
                fill: "#4a9cc0",
                strokeWidth: number === 3 ? 3 : 0,
                r: number === 3 ? 5 : 0,
              }}
              onMouseOver={() => {
                setNumber(3);
              }}
              strokeWidth={
                selected.includes("Neo") || selected.includes("All") ? 1 : 0
              }
              type="monotone"
              dataKey="neo"
              stroke="#4a9cc0"
              fillOpacity={1}
              fill={
                selected.includes("Neo") || selected.includes("All")
                  ? `url(#Blue2)`
                  : "none"
              }
            />

            <Area
              activeDot={{
                stroke: "#fff",
                fill: "#a075b9",
                strokeWidth: number === 4 ? 3 : 0,
                r: number === 4 ? 5 : 0,
              }}
              onMouseOver={() => {
                setNumber(4);
              }}
              strokeWidth={
                selected.includes("Ripple") || selected.includes("All") ? 1 : 0
              }
              type="monotone"
              dataKey="ripple"
              stroke="#a075b9"
              fillOpacity={1}
              fill={
                selected.includes("Ripple") || selected.includes("All")
                  ? `url(#Pink)`
                  : "none"
              }
            />
          </AreaChart>
        </ResponsiveContainer>
        <Grid />
      </div>
    </div>
  );
}

export default Graph;
