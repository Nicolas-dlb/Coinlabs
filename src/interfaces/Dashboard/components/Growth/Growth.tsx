/* eslint-disable react-hooks/exhaustive-deps */
import Graph from 'components/Graph/Graph';
import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { selectBalanceHistory, selectCryptoHistory, selectExpensesHistory, selectIncomeHistory } from 'redux/reducers/walletsSlice';
import './Growth.scss';
import Select from "components/Select/Select";

function Growth() {
    const [active, setActive] = useState("1m");
const balanceHistory = useSelector(selectBalanceHistory);
const [data, setData] = useState([{tokens: 0, Price: 0, timestamp: 3227632}])
 const [growthSelected, setGrowthSelected] = useState("Balance");
 const incomeHistory = useSelector(selectIncomeHistory);
 const cryptoHistory = useSelector(selectCryptoHistory);
 const expensesHistory = useSelector(selectExpensesHistory);

 const timeline = growthSelected === "Balance" ? balanceHistory : growthSelected === "Income" ? incomeHistory : growthSelected === "Expenses" ? expensesHistory : cryptoHistory;
  useEffect(() => {
   if (timeline) {
       
        setData([...Array(101).keys()].map((x) => ({
    tokens: x,
    Price: timeline.slice(-101)[x]?.total,
    timestamp: timeline.slice(-101)[x]?.timestamp
  })));
   }
  }, [timeline])

    return (
        <div  className="growth">
            <div className="tabs">
                <div className="tabs_left">
                    <span onClick={() => setActive("1w")} className={active === "1w" ? "btn_growth btn_growth_active" : "btn_growth"}>1w</span>
                <span onClick={() => setActive("1m")} className={active === "1m" ? "btn_growth btn_growth_active" : "btn_growth"}>1m</span>
                <span onClick={() => setActive("3m")} className={active === "3m" ? "btn_growth btn_growth_active" : "btn_growth"}>3m</span>
                <span onClick={() => setActive("6m")} className={active === "6m" ? "btn_growth btn_growth_active" : "btn_growth"}>6m</span>
                <span onClick={() => setActive("1y")} className={active === "1y" ? "btn_growth btn_growth_active" : "btn_growth"}>1y</span>
                </div>
                <div className="tabs_right">
                   
                   <Select items="currency" setCurrencySelected={setGrowthSelected} />
                </div>
            </div>
            <Graph data={data} color="Green" width={0} height={180} tooltip/>
        </div>
    )
}

export default Growth
