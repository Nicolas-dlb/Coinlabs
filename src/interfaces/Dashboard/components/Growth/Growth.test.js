/* eslint-disable no-unused-vars */
import { mount, render } from "enzyme";
import React from "react";
import { Provider } from "react-redux";
import Graph from "components/Graph/Graph";
import store from "redux/store";
import { act, fireEvent, screen, waitForElement } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  setBalanceHistory,
  setExpensesHistory,
} from "redux/reducers/walletsSlice";
import Growth from "./Growth";

describe("Growth", () => {
  let GrowthComponent;
  const setState = jest.fn();
  const useStateMock = (state) => [state, setState];
  jest.spyOn(React, "useState").mockImplementation(useStateMock);

  beforeEach(async () => {
    await act(async () => {
      GrowthComponent = mount(
        <Provider store={store}>
          <Growth />
        </Provider>
      );
    });
  });

  it("should render correctly", () => {
    expect(GrowthComponent).toMatchSnapshot();
  });

  it("should render Graph component", () => {
    expect(GrowthComponent.find(".graph").exists()).toBeTruthy();
  });

  // it("should render correct timeline", () => {
  //   expect(GrowthComponent.find("Graph").props().data).toStrictEqual([
  //     { tokens: 0, Price: 0, timestamp: 0 },
  //   ]);
  // });
});
