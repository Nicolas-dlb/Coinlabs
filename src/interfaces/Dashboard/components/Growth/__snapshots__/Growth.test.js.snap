// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Growth should render correctly 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Growth>
    <div
      className="growth"
    >
      <div
        className="tabs"
      >
        <div
          className="tabs_left"
        >
          <span
            className="btn_growth"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            1w
          </span>
          <span
            className="btn_growth btn_growth_active"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            1m
          </span>
          <span
            className="btn_growth"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            3m
          </span>
          <span
            className="btn_growth"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            6m
          </span>
          <span
            className="btn_growth"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            1y
          </span>
        </div>
        <div
          className="tabs_right"
        >
          <Select
            items="currency"
            setCurrencySelected={[Function]}
          >
            <div>
              <select
                className="currency exchange_element"
              >
                <option
                  className="option exchange_element"
                  id="growth_balance"
                  label="Balance"
                  value="Balance"
                >
                  Balance
                </option>
                <option
                  className="option exchange_element"
                  id="growth_income"
                  label="Income"
                  value="Income"
                >
                  Income
                </option>
                <option
                  className="option exchange_element"
                  id="growth_Expenses"
                  label="Expenses"
                  value="Expenses"
                >
                  Expenses
                </option>
                <option
                  className="option exchange_element"
                  id="growth_Crypto"
                  label="Crypto"
                  value="Crypto"
                >
                  Crypto
                </option>
              </select>
              <div
                className="currency-select exchange_element"
              >
                <svg
                  className="select-currency-arrow exchange_element"
                  fill="#605b95"
                  height="24px"
                  viewBox="0 0 24 24"
                  width="24px"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 0h24v24H0V0z"
                    fill="none"
                  />
                  <path
                    d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                  />
                </svg>
                <button
                  aria-label="btn-select"
                  className="btn-currency exchange_element"
                  id="btn-select"
                  type="button"
                  value="All"
                />
                <div
                  className="currency-list exchange_element"
                >
                  <ul
                    id="currency-items"
                  />
                </div>
              </div>
            </div>
          </Select>
        </div>
      </div>
      <Graph
        color="Green"
        data={
          Array [
            Object {
              "Price": 0,
              "timestamp": 0,
              "tokens": 0,
            },
          ]
        }
        height={180}
        tooltip={true}
        width={0}
      >
        <div
          className="graph_container"
        >
          <div
            className="graph"
          >
            <ResponsiveContainer
              debounce={0}
              height={180}
              width="90%"
            >
              <ResizeDetector
                handleHeight={true}
                handleWidth={true}
                onResize={[Function]}
                targetRef={
                  Object {
                    "current": <div
                      class="recharts-responsive-container"
                      style="width: 90%; height: 180px;"
                    />,
                  }
                }
              >
                <div
                  className="recharts-responsive-container"
                  style={
                    Object {
                      "height": 180,
                      "maxHeight": undefined,
                      "minHeight": undefined,
                      "minWidth": undefined,
                      "width": "90%",
                    }
                  }
                />
              </ResizeDetector>
            </ResponsiveContainer>
            <Grid>
              <div
                className="grid"
              >
                <span
                  className="grid_card"
                />
                <span
                  className="grid_card"
                />
                <span
                  className="grid_card"
                />
                <span
                  className="grid_card"
                />
                <span
                  className="grid_card"
                />
              </div>
            </Grid>
          </div>
        </div>
      </Graph>
    </div>
  </Growth>
</Provider>
`;
